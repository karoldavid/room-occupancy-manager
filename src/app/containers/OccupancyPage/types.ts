export enum RoomType {
  PREMIUM = 'premium',
  ECONOMY = 'economy',
}

export enum GuestsErrorType {
  RESPONSE_ERROR = 1,
  GUESTS_NOT_FOUND = 2,
}

export interface AvailableRooms {
  [name: string]: number;
}

export interface AvailableRoomsFormValues {
  [name: string]: {
    value: number;
  };
}

export interface Revenue {
  free: number;
  usage: number;
  total: number;
  currency: string;
}

export type RevenueRoomType = {
  [key in RoomType]: Revenue;
};

export interface BookingState {
  availableRooms: AvailableRooms;
  guests: number[];
}

export interface OccupancyState extends BookingState {
  loading: boolean;
  error?: GuestsErrorType | null;
}

export type ContainerState = OccupancyState;
