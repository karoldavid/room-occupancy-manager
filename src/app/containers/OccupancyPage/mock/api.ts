import data from './guests.json';

export interface MockResponse {
  data: number[];
}

const mockFetch = (
  success: boolean,
  timeout: number,
  throwError: boolean,
): Promise<MockResponse> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (!throwError && success) {
        const response: MockResponse = {
          data,
        };
        return resolve(response);
      } else {
        return reject({ status: 500, message: 'failed to fetch' });
      }
    }, timeout);
  });
};

export const fetchGuests = async (throwError = false) => {
  try {
    const response: MockResponse = await mockFetch(true, 1000, throwError);
    return response;
  } catch (e) {
    console.log(e.message);
    throw e;
  }
};
