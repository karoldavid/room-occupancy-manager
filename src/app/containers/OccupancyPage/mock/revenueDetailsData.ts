import { RoomType, RevenueRoomType } from '../types';

export const revenueDetailsData: RevenueRoomType = {
  [RoomType.PREMIUM]: { free: 0, usage: 0, total: 0, currency: 'EURO' },
  [RoomType.ECONOMY]: { free: 0, usage: 0, total: 0, currency: 'EURO' },
};
