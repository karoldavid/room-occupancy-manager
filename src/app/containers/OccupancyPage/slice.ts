import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import {
  ContainerState,
  AvailableRooms,
  RoomType,
  GuestsErrorType,
} from './types';

export const initialState: ContainerState = {
  availableRooms: { [RoomType.PREMIUM]: 0, [RoomType.ECONOMY]: 0 },
  guests: [],
  loading: false,
  error: null,
};

const occupancySlice = createSlice({
  name: 'occupancy',
  initialState,
  reducers: {
    updateAvailableRoomsForm(state, action: PayloadAction<AvailableRooms>) {
      state.availableRooms = action.payload;
    },
    loadGuests(state) {
      state.guests = [];
      state.loading = true;
      state.error = null;
    },
    guestsLoaded(state, action: PayloadAction<number[]>) {
      state.guests = action.payload;
      state.loading = false;
      state.error = null;
    },
    guestsError(state, action: PayloadAction<GuestsErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = occupancySlice;
