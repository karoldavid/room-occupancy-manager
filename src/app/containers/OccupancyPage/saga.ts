import { call, put, takeLatest } from 'redux-saga/effects';
import { actions } from './slice';
import { GuestsErrorType } from './types';
import { fetchGuests, MockResponse } from './mock/api';

export function* getGuests() {
  try {
    const response: MockResponse = yield call(fetchGuests);
    if (response && response.data) {
      yield put(actions.guestsLoaded(response.data));
    }
  } catch (err) {
    // eslint-disable-next-line no-console
    console.log(err);

    if (err.response?.status === 404) {
      yield put(actions.guestsError(GuestsErrorType.GUESTS_NOT_FOUND));
    } else {
      yield put(actions.guestsError(GuestsErrorType.RESPONSE_ERROR));
    }
  }
}

export function* occupancySaga() {
  yield takeLatest(actions.loadGuests.type, getGuests);
}
