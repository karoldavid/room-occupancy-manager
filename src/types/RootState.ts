import { OccupancyState } from 'app/containers/OccupancyPage/types';

export interface RootState {
  occupancy: OccupancyState;
}
