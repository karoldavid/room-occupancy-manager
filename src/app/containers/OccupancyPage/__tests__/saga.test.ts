import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';
import { occupancySaga, getGuests } from '../saga';
import { MockResponse } from '../mock/api';
import data from '../mock/guests.json';
import { GuestsErrorType } from '../types';

describe('getGuests Saga', () => {
  const response: MockResponse = {
    data,
  };
  let getGuestsIterator: ReturnType<typeof getGuests>;

  beforeEach(() => {
    getGuestsIterator = getGuests();

    const selectDescriptor = getGuestsIterator.next(response).value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the guestsLoaded action if it requests the data successfully', () => {
    const putDescriptor = getGuestsIterator.next(response).value;
    expect(putDescriptor).toEqual(
      put(slice.actions.guestsLoaded(response.data)),
    );
  });

  it('should dispatch the response error', () => {
    const requestDescriptor = getGuestsIterator.next(response).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getGuestsIterator.throw(new Error('some error'))
      .value;
    expect(putDescriptor).toEqual(
      put(slice.actions.guestsError(GuestsErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('occupancySaga Saga', () => {
  const guestsIterator = occupancySaga();
  it('should start task to watch for loadGuests action', () => {
    const takeLatestDescriptor = guestsIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.actions.loadGuests.type, getGuests),
    );
  });
});
