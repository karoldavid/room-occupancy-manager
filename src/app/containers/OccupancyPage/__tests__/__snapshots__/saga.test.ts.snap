// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGuests Saga should dispatch the guestsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getGuests Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getGuests Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Array [
        23,
        45,
        155,
        374,
        22,
        99,
        100,
        101,
        115,
        209,
      ],
      "type": "occupancy/guestsLoaded",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
